racingcar 미션

pkalsh opened the pull request.

oereo committed:
docs(guide): 구현할 기능목록과 예외처리 목록 추가

oereo committed:
test: Car의 move 메서드 호출시 한칸 전진하는 테스트 작성

oereo committed:
feat: 자동차가 움직이는 기능 구현

oereo committed:
test: 전진 여부 판단 test code 작성

oereo committed:
feat: 전진 여부 판단 기능 구현

oereo committed:
feat: 0~9 범위의 난수 생성 기능 구현

oereo committed:
test: car 객체를 CarManager에 추가하는 test 작성

oereo committed:
feat: Car 객체 추가 테스트 통과하는 기능 구현

- 생성된 Car 객체를 보관하는 구현체로는 인메모리 List로 구현
- add method: 관리하는 car 객체 추가
- size method: 관리하고 있는 car 객체 개수 반환

oereo committed:
test: 중복되는 이름의 Car 객체를 추가할 시 NotDulicateNameException를 던지는 test 작성

oereo committed:
fix: 독립된 테스트케이스 환경 구성

oereo committed:
feat: 중복되는 이름의 Car 객체를 검사하고 존재하면 NotDuplicateNameException 던지는 기능 구현

oereo committed:
test: Car 객체의 상태를 출력 포맷에 맞게 String으로 반환하는 test 작성

oereo committed:
feat: Car 객체의 상태를 출력 포맷에 맞게 String으로 반환하는 기능 구현

oereo committed:
test: 빈 String의 이름을 가진 Car 객체를 생성하면 NotBlankException을 던지는 test 작성

oereo committed:
feat: 빈 String의 이름을 가진 Car 객체를 생성하면 NotBlankException을 던지는 기능 구현

oereo committed:
refactor: 루트 패키지 racingcar 추가

oereo committed:
refactor: Discriminator 패키지 위치 변경

oereo committed:
refactor: 플레이어 추가를 addAllCars method 로 test case 작성

oereo committed:
refactor: 플레이어 추가를 addAllCars method 로 구현

- add(Car) 메서드 삭제
- addAllCars(List<Car>)로 제한

oereo committed:
test: 랜덤 숫자 리스트 기준에 따라 플레이어가 이동하는 test case 작성

oereo committed:
feat: 랜덤 숫자 리스트 기준에 따라 플레이어가 이동하는 method 구현

oereo committed:
refactor: car 객체와 생성된 난수를 dto로 mapping

Before:
- List로 생성된 난수를 참조하여 일치하는 인덱스의 car 객체를 움직임

After:
- CarNumberDto로 Car 객체와 난수를 묶어 인덱스 참조 제거

oereo committed:
test: dto 리스틀를 생성하는 test case 작성

- dto에 존재하는 car 객체가 CarManager의 collection에 존재하는 car 객체가 같은지 여부 파악하는 test case 추가

oereo committed:
feat: dto에 car 객체와 난수 생성하는 기능 구현

oereo committed:
feat: 출력 담당 Printer class 구현

oereo committed:
feat: 전체 car 의 state 를 출력하는 printCarState method 구현

oereo committed:
feat: 사용자의 입력을 받는 역할을 담당하는 Receiver class 구현

oereo committed:
test: 승자가 한명일 때 그 이름을 출력하는 test case 작성

oereo committed:
test: 승자가 여러명일 때 그 이름들을 comma로 구분하여 출력하는 test case 작성

oereo committed:
feat: 승자를 구하고 그 승자이름들을 comma로 구분하여 출력하는 기능 구현

oereo committed:
feat: 게임 흐름 제어하는 RacingGameApplication 구현

oereo committed:
fix: 정상적이지 않은 입력값에 대한 예외 처리 NotBlankException

Jummi10 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private void checkBlankName(String name) {
+        if (name.equals("")) {
```
`equals("")` 대신 `isEmpty()`나 `isBlank()`를 사용해보는건 어떨까요!
[참고](https://thisisnew-storage.tistory.com/3)

pkalsh replied:
오!

Jummi10 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public String toString() {
+        return name + " : " + "-".repeat(Math.max(0, position));
+    }
```
오호 for문을 사용하지 않고 position을 출력할 수 있군요! 배워갑니다 ㅎㅎ

Jummi10 commented on src/main/java/racingcar/exception/NotZeroRoundException.java:
```diff
@@ -0,0 +1,9 @@
+    public NotZeroRoundException() {
+        super(NOT_ZERO_ROUND_EXCEPTION_MESSAGE);
+    }
+}
```
`No newline at end of file`이 떴군요~

Jummi10 commented on src/main/java/racingcar/util/RandomGenerator.java:
```diff
@@ -0,0 +1,16 @@
+    private static final Random RANDOM = new Random();
+
+    private RandomGenerator(){
+
```
불필요한 공백이 들어갔네요!

Jummi10 commented on src/main/java/racingcar/util/Discriminator.java:
```diff
@@ -0,0 +1,9 @@
+public class Discriminator {
+    private static final int THRESHOLD = 4;
+
+    public static boolean isMove(int randomNumber) {
```
`randomNumber`가 주어졌을 때 move할 수 있느냐 없느냐이므로 `canMove()`라고 명명하는 것은 어떤가요~?

pkalsh replied:
괜찮군요 👍

Jummi10 commented on src/main/java/racingcar/ui/Printer.java:
```diff
@@ -0,0 +1,44 @@
+        receiver.clearBuffer();
+        System.out.println(INPUT_MISMATCH_EXCEPTION_MESSAGE);
+    }
+}
```
여기에도 no newline at end of file이 떴군요!

Jummi10 commented on src/main/java/racingcar/ui/Receiver.java:
```diff
@@ -0,0 +1,21 @@
+        return scanner.nextInt();
+    }
+
+    public void clearBuffer() {
+        if (scanner.hasNextLine()) {
+            scanner.nextLine();
+        }
+    }
```
`hasNextLine()`을 사용한 이유가 있나요?? `scanner.nextLine()`을 `clearBuffer`라고 한 이유는 무엇인가요! 그저 궁금해서 묻습니다 ㅎㅎ

pkalsh replied:
nextInt 메서드 호출로 숫자와 다른 입력이 들어오면 예외를 발생시키고 읽지 못했던 나머지 문자열들이 버퍼에 남게 됩니다.
이런 경우 다음  입력을 받을때 똑같은 예외를 발생시켜 의도치 않은 오류가 발생시킵니다.
그런 경우를 대비해 입력 버퍼에 데이터가 남아있다면 nextLine 해줬습니다!

Jummi10 commented on src/main/java/racingcar/repository/CarManager.java:
```diff
@@ -0,0 +1,88 @@
+        return winnerNames.toString();
+    }
+
+    private List<Car> winnerList() {
```
메서드명을 get이나 make 등을 사용해서 바꾸는건 어떨까요!

Jummi10 commented on src/main/java/racingcar/RacingGameApplication.java:
```diff
@@ -0,0 +1,94 @@
+            carList.add(new Car(name));
+        }
+        return carList;
+    }
```
차를 만드는 메서드이므로 `CarManager`의 메서드여도 괜찮을 것 같다는 생각이 드네요

Jummi10 commented on src/main/java/racingcar/RacingGameApplication.java:
```diff
@@ -0,0 +1,94 @@
+        String line = receiver.receiveCarNames();
+        try {
+            List<Car> carList = createAllCars(splitInputLine(line));
+            carManager.addAllCars(carList);
```
> List 대신 다른 자료형(Set…)을 써야 하는 경우가 오면 어떻게 해야 할까? 기존 변수명이 적절한 의미를 나타내지 못하게 되므로 결국 변수명을 변경해야 하는 번거로움이 생긴다. 변수 이름에 자료형을 쓰지 않아도 타입을 통해 충분히 어떤 변수인지 파악이 가능하다.
List, Collection 등의 자료형은 복수형으로 표현하는 것이 좋다.
[참고](https://woowacourse.github.io/javable/post/2020-04-24-variable_naming/)

`carList`가 좋은 변수명인것 같진 않아 보이네요!

Jummi10 commented on src/main/java/racingcar/RacingGameApplication.java:
```diff
@@ -0,0 +1,94 @@
+        return carList;
+    }
+
+    private int receiveRounds() {
+        int rounds = 0;
+        boolean isInvalidInput = true;
+        while (isInvalidInput) {
+            try {
+                rounds = receiver.receiveNumberOfRounds();
+                throwExceptionIfInputIsZero(rounds);
+                isInvalidInput = false;
+            } catch (InputMismatchException e) {
+                printer.printInputMismatchExceptionMessage(receiver);
+            } catch (IllegalArgumentException e) {
+                printer.printExceptionMessage(e);
+            }
+        }
+        return rounds;
+    }
```
`receiver.receiveNumberOfRounds()`에서 round 입력을 받아와서 valid한지 아닌지 검증하는 기능인 것 같은데
그렇다고 메서드명을 `validateRounds()` 등을 사용할 경우 round를 반환하는 것도 올바르다고 생각되지는 않지만, 이 메서드의 기능에 대해서 생각하자면 `receiveRounds()`도 괜찮은 이름인지 의문이 드네요
@pkalsh 님이나 다른 리뷰어님은 어떻게 생각하시는지 궁금합니다!

pkalsh replied:
결국 입력 받은 값을 int로 변환하여 내놓기 때문에 receiveRounds로 가져가도 되지 싶습니다.
완전히 엎은 코드에서도 이 네이밍 밖에 생각나지 않더군요.. 😢

Jummi10 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,46 @@
+        car.move();
+
+        //when
+        String expected = car.toString();
+
+        //then
+        assertThat(expected).isEqualTo("name : ---");
```
`Assertions.assertThat()`의 parameter를 보면 `assertThat(actual).isEqualTo(expected)`라고 뜨네요
인텔리제이 단축키 
> ⌘P(Ctrl+P) : 현재 위치한 메서드의 파라미터 정보를 조회합니다. (Parameter info -within method call arguments)

로 확인하실 수 있습니다 ㅎㅎ

![image](https://user-images.githubusercontent.com/55528172/111033063-c27ba580-8452-11eb-9403-2139dc98583d.png)

Jummi10 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,46 @@
+        assertThat(expected).isEqualTo("name : ---");
+    }
+
+    @Test
+    void 빈문자열의_이름이_주어지면_NotBlankException을_던진다() {
+        //then
+        assertThatExceptionOfType(NotBlankException.class)
+                .isThrownBy(() -> new Car(""));
+    }
```
`.withMessage("%s", "빈 문자열 이름의 차를 생성할 수는 없습니다.")` 나
`.withMessageContaining("빈 문자열 이름의 차를 생성할 수는 없습니다.")`로 한 번 더 확인해주어도 좋을 것 같아요

pkalsh replied:
적용했습니다~

Jummi10 reviewed (changes requested):
리뷰가 조금 늦었습니다.. 여러 경우에 대해서 테스트 케이스까지 잘 작성해주신 것 같아요
너무 수고 많으셨습니다!! :)

kouz95 commented on src/main/java/racingcar/RacingGameApplication.java:
```diff
@@ -0,0 +1,94 @@
+
+    private List<Car> createAllCars(List<String> carNames) {
+        List<Car> carList = new ArrayList<>();
+        if (carNames.size() == 0) {
```
`0`도 상수화하여 의미를 담을수 있을것 같아요 🙂

kouz95 commented on src/main/java/racingcar/repository/CarManager.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CarManager {
```
클래스 이름을 `~er` 또는 `~or`로 만들지 말라는 얘기가 있습니다.
`Manager` 라는 것은 그 자체로 많은 책임을 질 수 있는 SuperClass가 될 여지가 있습니다.

pkalsh replied:
라운드를 진행하는 책임만 가지는 것을 의미하는 이름인 RoundProcessor 같은 경우에도 피해야할 경우일까요?
~er, ~or을 제외하고 이름 짓기가 너무 힘드네요 😢

kouz95 commented on src/main/java/racingcar/repository/CarManager.java:
```diff
@@ -0,0 +1,88 @@
+        return cars.size();
+    }
+
+    public void addAllCars(List<Car> cars) {
+        for (Car car : cars) {
+            checkDuplicateName(car);
+            this.cars.add(car);
+        }
+    }
```
생성자에서 List를 주입받아 생성시키는것에 대해선 어떻게 생각하시나요?

pkalsh replied:
마침 누워서 일급 컬렉션에 대해 고민하다가 이 생각을 하고 있었는데 불변 객체로 만들기 위해서는 add를 막는 것이 필수인 듯 하네요.

kouz95 commented on src/main/java/racingcar/RacingGameApplication.java:
```diff
@@ -0,0 +1,94 @@
+        return carList;
+    }
+
+    private int receiveRounds() {
+        int rounds = 0;
+        boolean isInvalidInput = true;
+        while (isInvalidInput) {
+            try {
+                rounds = receiver.receiveNumberOfRounds();
+                throwExceptionIfInputIsZero(rounds);
+                isInvalidInput = false;
+            } catch (InputMismatchException e) {
+                printer.printInputMismatchExceptionMessage(receiver);
+            } catch (IllegalArgumentException e) {
+                printer.printExceptionMessage(e);
+            }
+        }
+        return rounds;
+    }
```
만약 `Round`라는 객체가 생긴다면 Application의 책임을 좀 더 분리할 수 있을것 같아요 🙂

pkalsh replied:
적용했습니다~

kouz95 commented on src/main/java/racingcar/RacingGameApplication.java:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    private void proceedRound(int rounds) {
+        printer.printResultHeader();
+        for (int i = 0; i < rounds; i++) {
+            List<CarNumberDto> carNumberDtos = carManager.generateCarNumberDtos();
+            carManager.moveAllCars(carNumberDtos);
+            carManager.printCarState(printer);
+        }
+    }
```
DTO를 사용한 이유는 무엇인가요?
일반적으로 DTO는 말그대로 데이터를 데이터를 전달하기위해 (Data Transfer Object) 사용합니다.

View와 Domain을 분리하기위해 사용되기도 하는데, Domain에서 만들어낸 결과 또는 Domain 객체의 상태를 담아 넘겨주곤 합니다. 이 경우엔 View와 관련된 부분이 Domain을 직접 의존하는것을 피할 수 있게 됩니다.

위 코드에서 DTO는 CarManager에서 CarManager로 이동하는것처럼 보여져서 질문드립니다!

pkalsh replied:
세션 중에 DTO 패턴 힌트를 주셔서 최대한 그 의미를 끼워 맞추다가 네이밍 됐습니다.
문제는 CarManager 내의 컬렉션의 크기를 외부에 드러내지 않고 그 크기만큼 난수를 생성하는 것에 어려움을 겪어서 이런 코드가 생산됐습니다 ㅠ

kouz95 commented on src/main/java/racingcar/repository/CarManager.java:
```diff
@@ -0,0 +1,88 @@
+package racingcar.repository;
```
CarManager를 repository 패키지에 둔 이유가 궁금하군요!

pkalsh replied:
Car들을 담고 있다해서 짜낸 의미인데 일급 컬렉션을 알게된 이후로 바보 같은 네이밍이었다는 걸 깨달았습니다..

kouz95 commented on src/main/java/racingcar/repository/CarManager.java:
```diff
@@ -0,0 +1,88 @@
+                .orElseThrow(IllegalArgumentException::new);
+
+        return cars.stream()
+                .filter(car -> car.getPosition() == winnerPosition)
```
Car에게 메세지를 보내볼까요?

pkalsh replied:
적용했습니다~

kouz95 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public void move() {
+        position += 1;
```
`1`도 상수화 할 수 있지 않을까요?

kouz95 commented on src/main/java/racingcar/repository/CarManager.java:
```diff
@@ -0,0 +1,88 @@
+
+        for (int i = 1; i < winners.size(); i++) {
+            Car winner = winners.get(i);
+            winnerNames.append(", ");
```
상수화!

kouz95 reviewed (changes requested):
고생하셨습니다 🙂

스트림도 잘 활용하시는군요! 👍
리뷰 확인해주세요!

pkalsh committed:
refactor: StringJoiner로 우승자 메시지 생성 구현 코드 축소

pkalsh committed:
refactor: 공백 입력 검증 기능, String api 활용

pkalsh committed:
chore: no newline at end of file 해결

pkalsh committed:
refactor: RacingGameApplication의 책임 축소

Before:
- 게임의 순서 보장
- 이 클래스 내에 절차 지향적인 코드가 집중되어 있었음
- 다른 객체에게 명령하기 보다 메시지를 보내서 이 객체의 역할을 다 하는데에만 집중

After:
- 게임 한판 실행후 우승자 목록 출력

pkalsh committed:
refactor: 책임 재분배를 위한 클래스 정리

pkalsh committed:
refactor: Car 객체의 이름 타입을 Name 객체로 변경

pkalsh committed:
refactor: 이름 객체 정의

- 이름에 관한 제약 조건을 이 객체 안에서 검증

pkalsh committed:
test: 이름과 관련한 예외 사항 테스트 검증

pkalsh committed:
refactor: Cars 객체 이름 변경

pkalsh committed:
refactor: Car 객체 이름 입력을 파싱해서 생성하는 책임을 CarFactory로 분리

pkalsh committed:
refactor: Round의 컬렉션을 래핑하는 Rounds 일급 컬렉션 구현

- 실제 Round 리스트를 생성하는 책임은 RoundFactory에게

pkalsh committed:
refactor: 입력으로 인한 예외 처리 책임을 Receiver의 필터 체인에 할당

- InputExceptionHandler와 InputValidator를 데코레이터 패턴으로 Receiver의 필터 구현
- InputValidator 다음에 InputExceptionHandler가 꼭 위치해야 하므로 InputExceptionHandler의 생성자 인자 타입을 InputValidator로 제한

pkalsh committed:
test: 입력 관련 테스트 작성

pkalsh committed:
fix: Car 객체 동등성 조건 수정

- Name이 객체이므로 그 값을 비교

pkalsh committed:
fix: 팩토리 객체들이 던지는 예외 처리 구현

pkalsh committed:
refactor: 새로운 설계에 맞게 Printer 객체 구현

pkalsh committed:
chore: 문서 업데이트, 자바 버전 설정

pkalsh committed:
test: 팩토리 객체 테스트 작성

- CarFactory
- RoundFactory

pkalsh committed:
refactor: 우승자들을 담기 위한 객체 Winner 구현

pkalsh committed:
fix: Round 리스트 안의 객체가 null인 버그 수정

pkalsh committed:
refactor: 매개변수 이름 변경

pkalsh committed:
fix: InputMismatchException에 대한 에러 메시지 출력되지 않는 버그 수정

pkalsh committed:
refactor: Car 객체에 일급 컬렉션 적용

pkalsh committed:
refactor: 경주 게임에 참여하는 Car 객체들과 Rounds 객체들로 경주를 진행하는 Racing 객체 구현

pkalsh committed:
refactor: 움직이는 비즈니스 로직 전략 패턴 적용

- Round 객체 내에서 실제 Car 객체들의 상태 변경 요청
- MoveStrategy DI
- 하지만 Cars를 테스트하기엔 좋지 않은 구조

kouz95 commented on src/main/java/racingcar/ui/receiver/InputValidator.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.regex.Pattern;
+
+public class InputValidator extends ReceiverFilter {
```
UI 로직과 비지니스 로직을 분리하여야 합니다.
`자동차 이름의 입력은 콤마로 구분하여 하나 이상 입력해주세요.`는 비지니스 로직일까요 UI 로직 일까요?

현재, UI의 환경이 바뀌더라도 Receiver의 다형성을 통해 InputValidator를 적용하면 `자동차 이름의 입력은 콤마로 구분하여 하나 이상 입력해주세요.`는 해결 할 수 있을 것이라고 생각됩니다. (이것을 의도 하셨나요?)

하지만, 실질적인 비지니스 로직을 담당하는 도메인 객체인 `Factory`는 `carNames`로 `"a,"` 가 주어졌을때 `"a"` 와 `""`의 이름을 가진 Car 생성을 시도합니다. 이후 `Name` 클래스로 인해 생성에 실패하게 됩니다.

이 상황을 중복으로 검증을 하고 있다고 생각할 수도 있지 않을까요? 물론, 중복으로 검증하는게 나쁜건 아닐겁니다. 
근본적으로 왜 UI로직과 비지니스 로직을 분리해야 되는 걸까요?
UI가 도메인 로직에 관심을 갖게 되면 어떤 일이 일어날까요?

pkalsh replied:
isValidPattern이 Name 클래스의 Blank 검증 로직과 중복된다는 것을 발견하지 못했었네요.
아마 입력 문자열에 관한 검증 로직만을 생각하다 일어난 결과 같습니다.

UI 로직과 비즈니스 로직을 분리해야 하는 이유는 
우선, 자신이 담당하고 있는 책임만을 다루고 있어야 응집도가 높아지고 객체간 결합도를 낮출 수 있기 때문이라 생각합니다.
응집도를 높이고, UI와 도메인의 결합도를 낮춰야 할 이유는 MVC 패턴에서도 다루고 있습니다.
UI는 자주 변경될 사항이고ttern이 Name 클래스의 Blank 검증 로직과 중복된다는 것을 발견하지 못했었네요.
아마 입력 문자열에 관한 검증 로직만을 생각하다 일어난 결과 같습니다.

UI 로직과 비즈니스 로직을 분리해야 하는 이유는 
우선, 자신이 담당하고 있는 책임만을 다루고 있어야 응집도가 높아지고 객체간 결합도를 낮출 수 있기 때문이라 생각합니다.
응집도를 높이고, UI와 도메인 간의 결합도를 낮춰야 할 이유는 MVC 패턴에서도 다루고 있습니다.

도메인 객체들은 보통 비즈니스 자체가 흔들리지 않는 이상 변경될 이유가 없어야 합니다.
그만큼 안정적이기 때문에 다른 계층의 객체들이 알고 있어도 change propagation이 일어날 일이 거의 없습니다.
반면, UI 로직은 사용자들에 보여지는 부분이기 때문에 변경될 일이 많습니다.
따라서 MVC 패턴에서도 Model에서 View 컴포넌트로는 약하게 결합되어 있는 형태를 갖추고 있습니다.

 결국 질문 던져주신 UI 로직에서와 비즈니스 로직 중복으로 검증하는 경우, 비즈니스 로직 (자동차 이름의 입력을 콤마로 구분해 하나 이상 입력)이 변경되면 두 코드 모두 변경해주지 않았을 때 의도치 않은 결과가 나올 수 있어 디버깅이 어렵습니다.

 CarFactory 테스트에서 validateBlank에 대한 예외 테스트가 되지 않은 부분이 이것으로 설명이 되네요. UI 로직과 비즈니스 로직의 구분을 더 의도적으로 구분하여 생각할 수 있는 물음이 됐습니다. 😄

kouz95 commented on src/main/java/racingcar/domain/Racing.java:
```diff
@@ -0,0 +1,23 @@
+        this.rounds = rounds;
+    }
+
+    public Winner winners(Printer printer) {
+        rounds.process(cars, printer);
+        List<String> maxPositionCarNames = cars.getMaxPositionCarNames();
+        return new Winner(maxPositionCarNames);
```
현재 시스템의 도메인 객체와 ui객체에게 명령을 내리는 곳은 `RacingGameApplication`입니다.
MVC로 생각한다면 `RacingGameApplication`은 Controller에 해당할 것 입니다.

Printer가 도메인 객체에 들어오지 않도록 해보세요 🙂

pkalsh replied:
기능이 하나뿐인 컨트롤러를 생각하니 책임 분배가 명확해지네요 👍

kouz95 commented on src/main/java/racingcar/domain/Winner.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @Override
+    public String toString() {
+        StringJoiner stringJoiner = new StringJoiner(DELIMITER);
+        winnerNames.forEach(stringJoiner::add);
+        return stringJoiner.toString();
+    }
```
stream을 활용하여 join을 할 수도 있습니다.

[Java 8 Streams | Collectors.joining() method with Examples](https://www.geeksforgeeks.org/java-8-streams-collectors-joining-method-with-examples/)

pkalsh replied:
for문을 쓰지 않을 수 있겠네요 👍

kouz95 commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,47 @@
+        return position == winnerPosition;
+    }
+
+    @Override
+    public String toString() {
+        return name.getName() + " : " + "-".repeat(Math.max(0, position));
+    }
```
toString을 Override할땐 객체의 참조 값을 반환하는 것이 아닌, 객체의 상태를 반환하는 것이 용이하다고 판단 될때 입니다.

```java
@Override
    public String toString() {
        return "Car{" +
                "name=" + name +
                ", position=" + position +
                '}';
    }
```

물론, `System.out.println(car);` 만으로 해결되는 것이 매력적이긴 하겠지만, ui와 강한 결합을 갖게 됩니다.
`a : ---` 는 자동차의 이름과 위치를 표현하기 위한 ui의 정책이지 않나요?

pkalsh replied:
그러네요, 출력 형태가 바뀐다면 Car 도메인 코드를 변경하게 되니 UI와 강한 결합을 가진다는 말이 와닿습니다.
position을 외부에 노출시키지 않으려고 하다보니 이런 코드를 생산했습니다.
UI 객체에 책임을 분리시켜 보겠습니다.

kouz95 commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private void validate() {
+        cars.forEach(car -> {
+            long numberOfSameNameCars = cars.stream()
+                    .filter(car::equals)
+                    .count();
+
+            if (isMoreThanTwo(numberOfSameNameCars)) {
+                throw new IllegalArgumentException(DUPLICATED_NAME_EXCEPTION_MESSAGE);
+            }
+        });
```
```suggestion
        if (cars.size() != new HashSet<>(cars).size()) {
            throw new IllegalArgumentException(DUPLICATED_NAME_EXCEPTION_MESSAGE);
        }
        });
```

중복에 관련된 부분은 Set을 사용하여 처리할 수도 있습니다 🙂

pkalsh replied:
신박한 방법입니다... 😮

kouz95 commented on src/main/java/racingcar/domain/round/Rounds.java:
```diff
@@ -0,0 +1,26 @@
+        this.rounds = new ArrayList<>(rounds);
+    }
+
+    public void process(Cars cars, Printer printer) {
+        printer.printResultHeader();
+        rounds.forEach(round -> {
+            round.process(cars, new MoreThanFourMoveStrategy(new Random()));
+            printer.printCarState(cars);
+        });
+    }
```
Printer 분리!

pkalsh replied:
printer를 넘기기 보다 결과를 리턴하여 RacingGameApplication에서 출력하는 방법으로 바꿔봐야겠네요 👍

kouz95 reviewed (changes requested):
고생하셨습니다. 여러가지 시도를 많이 해보셨군요.

Receiver를 구현하신 방식도 인상 깊습니다. 디자인 패턴을 적용하신것 같은데 (데코레이터?) 다형성에 대한 이해나 자신감이 생겼을거라고 생각해요. 좋은 경험을 하셨네요. 🙂

리뷰 반영하시다가 막히는 부분이 있다면 언제든지 질문해주세요. (토론방에서 답변이 안된것은 세션 시간을 활용하여 답변 드리겠습니다. 개인적으로 연락해주셔도 좋습니다.)

그리고 테스트 커버리지를 한번 확인해주세요. ui가 아닌 도메인 객체들 (factory나 strategy도 도메인입니다)에 대해선 getter같은 것을 제외하곤 대부분 테스트 해야합니다. 

물론, ui를 테스트한건 대단합니다 👏 `MethodSource`, `ValueSource` 👍👍

pkalsh reviewed (commented):
테스트 커버리지가 채워지지 않은 이유는 CarsTest의 getMaxPositionCarNames나 moveEachCar 등이 Random이라는 외부 모듈과 강하게 결합되어 있어서 테스트할 방법이 없었네요..
이 결합을 약하게 하려는 시도를 해봤지만 어려워서 슬랙에도 도움을 청했습니다만 같은 고민을 하시는 분들이 안계시는 듯하더군요.
이 코드 푸시한 날은 지쳐서 포기하고 푸시했었는데 오늘 다시 한번 고민해보겠습니다.
빠른 코드 리뷰 감사합니다. 3따봉 드립니다 👍 👍 👍

pkalsh committed:
refactor: 중복 이름 검증 구현 방법 변경

- 중복은 Set으로 간단하게 구현 가능

pkalsh committed:
refactor: Car의 불변성 보장, 그 상태들을 값 객체로 변경

pkalsh committed:
refactor: Racing가 받는 메시지 세분화

- start : 주어진 이동 전략으로 각각의 차 객체들을 움직이고 모든 라운드의 결과를 반환
- winners : 마지막 라운드 결과에 대한 우승자를 Winner 객체로 반환

pkalsh committed:
refactor: StringJoiner에서 String.join으로 변경

pkalsh committed:
refactor: 이동 전략 객체들은 움직일 수 있는지만 판단하는 책임 부여

- 실제 움직이는 행위는 Car 객체에서 수행

pkalsh committed:
refactor: Round의 책임 확대

- 기존 Rounds와 RoundFactory가 지닌 책임을 Round가 수행
- 모든 라운드를 진행하고 결과를 반환하는 책임
- 라운드 입력을 검증하는 책임

pkalsh committed:
refactor: UI 로직을 모든 도메인 로직으로부터 분리

- UI를 사용할 수 있는 객체는 RacingGameApplication 뿐
- Cars 출력에 실제 Car 객체가 필요하므로 그 구현을 최대한 드러내지 않으면서 getter 메서드 구현 -> Collection

pkalsh committed:
fix: Comparable<Position> 오버라이드 메소드 구현 수정

- 오름차순 정렬로 변경

pkalsh committed:
refactor: VO의 getter도 필드 이름으로 네이밍

pkalsh committed:
refactor: 중복 이름 검증 책임 이동

- Cars -> CarFactory

pkalsh committed:
test: 중복 이름 검증 테스트 이동

- CarsTest -> CarFactoryTest

pkalsh committed:
test: toString 제거하면서 테스트도 제거

pkalsh committed:
refactor: 라운드 입력과 빈 문자열 입력 검증과 같은 비즈니스 로직 관련 검증 책임은 도메인 객체에게 맡김

- 콤마만 연속으로 입력되는 경우는 UI 로직이라 판단

pkalsh committed:
fix: 재귀 구조로 예외 처리해야 두번 이상의 비정상 입력시에도 대처 가능

pkalsh committed:
test: InputValidator에서 검증하는 문자열은 콤마만 연속하여 입력되는 경우로 한정하여 테스트 축소

pkalsh committed:
test: 콤마만 입력되는 경우는 InputValidator에서 검증하면서 CarFactoryTest에서 제거

pkalsh committed:
feat: 난수 생성 시작 숫자로 음수 사용시 예외

pkalsh committed:
test: 경계값 이동 전략 테스트 작성

pkalsh committed:
refactor: Car 객체의 불변성 보장

pkalsh committed:
fix: ",b,"와 같은 입력은 빈 문자열 검증에 의해 예외 처리되도록 수정

pkalsh committed:
test: 레이싱 테스트 작성

- 항상 움직이는 이동 전략으로 모든 메소드 테스트

pkalsh committed:
refactor: 라운드 입력 음수 검증은 Receiver 필터에서 하지 않기 때문에 catch문 제거

pkalsh committed:
test: 테스트 커버리지를 위한 테스트 추가

kouz95 commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,38 @@
+    private final Name name;
+    private final Position position;
+
+    public Car(String name) {
+        this.name = new Name(name);
+        this.position = new Position(0);
+    }
+
+    private Car(Name name, int position) {
+        this.name = name;
+        this.position = new Position(position);
+    }
```
저는 부생성자를 사용하게 되는 경우엔 모든 필드를 가진 생성자를 호출하도록 하고 있어요.

```suggestion
    private Car(Name name, Position position) {
        this.name = name;
        this.position = position;
        // 만약 유효성 검증이 필요하다면 이곳에서 진행합니다.
    }

    public Car(String name) {
        this(new Name(name), new Position(0));
    }
```

유효성 검사를 기본 생성자에서 진행하고, 모든 생성자가 그 생성자 메소드만 사용하게 한다면 객체가 어떤 생성자를 이용하더라도 유효성이 보장될거라고 생각해요 🙂

kouz95 commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,38 @@
+        return position;
+    }
+
+    public Car move(MoveStrategy moveStrategy) {
+        int nextPosition = position.getPosition();
+        if (moveStrategy.isMove()) {
+            nextPosition++;
+        }
+        return new Car(name, nextPosition);
+    }
```
```suggestion
    public Car move(MoveStrategy moveStrategy) {
        if (moveStrategy.isMove()) {
            return new Car(name, position.increase());
        }
        return this;
    }
```

```java
    public Position increase() {
        return new Position(position + INCREASE_SIZE);
    }
```

VO를 만들어내는 이점 중 하나는 내가 원하는 행동을 규정할 수 있다는 것입니다.
_get 하지말고 메세지를 보내라_ 에도 해당할 수 있겠네요 🙂

kouz95 commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,38 @@
+        return new Car(name, nextPosition);
+    }
+
+    public boolean isWinnerPosition(int winnerPosition) {
+        return position.getPosition() == winnerPosition;
+    }
```
getPosition().getPosition()이 뭔가 어색하지 않나요? 저는 이런식으로 구현했을것 같아요.

```suggestion
    public boolean isPositionEquals(Position position) {
        return this.position.equals(position);
    }
```

```java
public class Cars {

    ...

    public List<String> getMaxPositionCarNames() {
        Position max = cars.stream()
                .max(Comparator.comparing(Car::getPosition))
                .orElseThrow(() -> new IllegalStateException("cannot find, maybe no cars"))
                .getPosition();

        return cars.stream()
                .filter(car -> car.isPositionEquals(max))
                .map(car -> car.getName().getName())
                .collect(Collectors.toList());
    }
```

원시값을 포장한 값 객체(VO)는 자신이 만든 소프트웨어 세상의 소통 도구입니다. 불변 객체로 만들면 안전하니, 마음것 사용할 수도 있구요.

마찬가지로 `car.getName().getName()` 역시 `Winner` 가 `List<Name> winnerNames;` 를 가졌다면 안적혔을것 같아요 🙂

kouz95 commented on src/main/java/racingcar/domain/Winner.java:
```diff
@@ -0,0 +1,21 @@
+        this.winnerNames = winnerNames;
+    }
+
+    @Override
+    public String toString() {
+        return String.join(DELIMITER, winnerNames);
+    }
```
`toString()` 보단 delimiter를 인자로 받는 메소드를 만들어 `join` 기능을 지원할 수 도 있지 않을까요?
그렇게 만든다면 ui에 종속적이지도 않으면서 원하는 것을 해결할 수 있을것 같아요.

kouz95 commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,37 @@
+        return new Cars(cars);
+    }
+
+    public Collection<Car> getCollection() {
+        return cars;
+    }
```
`getCollection()`의 의도는 무엇이었나요?
만약, `List<Car>`를 주는건 뭔가 좀 위험할것 같고 마음이 썩 좋지 않다 였다면, 

```suggestion
    public List<Car> getCars() {
        return Collections.unmodifiableList(cars);
    }
```

이렇게도 만들수 있습니다.

kouz95 reviewed (approved):
요구사항 충분히 만족한것 같아 Approve 하겠습니다.
몇가지 도움이 될만한 부분들 리뷰 남겨놓았으니 확인해주세요 🙂
고생하셨습니다 👍
